QUICKFIX: gracefully handle unexpected unlinks

thousand opened the pull request.

rwjblue commented:
This seems good to me (if we get an `ENOENT` that means the file doesn't exist, so we should be "fine"), mind adding a test?

Also, can you describe how we get into the invalid state (just in case there are other issues that need to be addressed)?

rwjblue commented on lib/index.js:
```diff
@@ -13,7 +13,15 @@ var validateSortedUnique = util.validateSortedUnique;
+      fs.unlinkSync(outputPath);
+    } catch (e) {
+      // QUICKFIX: gracefully handle unexpected unlinks (from broccoli-dependency-funnel)
+      if (typeof e === 'object' && e !== null && e.code = 'ENOENT') {
```
This seems slightly incorrect, I believe you mean to not assign the `e.code` property?

```js
if (typeof e === 'object' && e !== null && e.code === 'ENOENT') { return; }
```

rwjblue replied:
This is what is failing the CI build.

rwjblue commented on lib/index.js:
```diff
@@ -13,7 +13,15 @@ var validateSortedUnique = util.validateSortedUnique;
+    try {
+      fs.unlinkSync(outputPath);
+    } catch (e) {
+      // QUICKFIX: gracefully handle unexpected unlinks (from broccoli-dependency-funnel)
```
Probably don't need this comment?

stefanpenner replied:
i wrote this :P

thousand commented:
Fixed the logic and added a test! Scenario that causes the issue is when the patch indicates an unlink, and the target directory doesn't contain that file. Test indicates the scenario we found where another process is deleting the file between the time the patch is generated and applied.
